pkg/services/promtypemigration/prom_mig.go

```diff
@@ -0,0 +1,89 @@
+package promtypemigration
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/repo"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/pluginstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type PromMigrationHandler interface {
+	Migrate(context.Context, *promMigrationService) error
+	getPrometheusDataSources(ctx context.Context, dataSourcesService datasources.DataSourceService) ([]*datasources.DataSource, error)
+}
+
+type dataSourceService interface {
+	DecryptedValues(ctx context.Context, ds *datasources.DataSource) (map[string]string, error)
+	UpdateDataSource(ctx context.Context, cmd *datasources.UpdateDataSourceCommand) (*datasources.DataSource, error)
+	GetDataSourcesByType(ctx context.Context, query *datasources.GetDataSourcesByTypeQuery) ([]*datasources.DataSource, error)
+}
+
+type promMigrationService struct {
+	cfg                *setting.Cfg
+	dataSourcesService dataSourceService
+	pluginStore        pluginstore.Store
+	pluginRepo         repo.Service
+	pluginInstaller    plugins.Installer
+}
+
+func (s *promMigrationService) applyMigration(ctx context.Context, pluginID string, promDataSources []*datasources.DataSource) error {
+	if len(promDataSources) == 0 {
+		return nil
+	}
+
+	// check to see if prom is installed, if not install it
+	if _, installed := s.pluginStore.Plugin(ctx, pluginID); !installed {
+		compatOpts := plugins.NewAddOpts(s.cfg.BuildVersion, runtime.GOOS, runtime.GOARCH, "")
+		err := s.pluginInstaller.Add(ctx, pluginID, "", compatOpts)
+		if err != nil {
+			return err
+		}
+	}
+
+	logger.Debug("performing prometheus data source type migration", "plugin", pluginID)
+
+	for _, ds := range promDataSources {
+		err := s.updateDataSourceType(ctx, ds, pluginID)
+		if err != nil {
+			return err
+		}
+	}
+
+	logger.Debug("prometheus data source type migration complete", "plugin", pluginID)
+
+	return nil
+}
+
+func (s *promMigrationService) updateDataSourceType(ctx context.Context, ds *datasources.DataSource, newType string) error {
+	secureJsonData, err := s.dataSourcesService.DecryptedValues(ctx, ds)
+	if err != nil {
+		return err
+	}
+	if ds.JsonData == nil {
+		return fmt.Errorf("no JsonData found for data source ID %d", ds.ID)
+	}
+	ds.JsonData.Set("prometheus-type-migration", true)
+	_, err = s.dataSourcesService.UpdateDataSource(ctx, &datasources.UpdateDataSourceCommand{
+		ID:             ds.ID,
+		Type:           newType,
```

Comment:
I wonder what happens for dashboards that use the old type (`prometheus`), do they still work even if that combination of `type + UID` does not match any data source?

Reply:
Yes because of the work Kevin did in this PR https://github.com/grafana/grafana/pull/103482